nes-network/network/src/sender/channel.rs

```diff
@@ -0,0 +1,387 @@
+                }
+            } else {
+                select! {
+                    response = Self::read_from_other_side(&self.cancellation_token, &mut self.reader) => self.handle_response(response?)?,
+                    request = Self::read_from_software(&self.cancellation_token, &mut self.queue) => self.handle_request(request?).await?,
+                    send_result = Self::send_pending(&self.cancellation_token, &mut self.writer, &mut self.pending_writes, &mut self.wait_for_ack) => send_result?,
+                }
+            }
+        }
+    }
+    async fn run(&mut self) -> Result<ChannelHandlerStatus> {
+        let status = match self
+            .run_internal()
+            .await
+            .expect_err("Handler should have not terminated by its own.")
+        {
+            ErrorOrStatus::Error(e) => return Err(e),
+            ErrorOrStatus::Status(status) => status,
+        };
+
+        // Fixup the ClosedBySoftwareButFailedToPropagate created in read_from_software
+        if let ChannelHandlerStatus::ClosedBySoftwareButFailedToPropagate(_) = status {
+            let Some(result) = self
+                .cancellation_token
+                .run_until_cancelled(self.writer.send(DataChannelRequest::Close))
+                .await
+            else {
+                return Ok(ChannelHandlerStatus::Cancelled);
+            };
+            if let Err(e) = result {
+                return Ok(ChannelHandlerStatus::ClosedBySoftwareButFailedToPropagate(
+                    e.into(),
+                ));
+            };
+            return Ok(ChannelHandlerStatus::ClosedBySoftware);
+        }
+
+        Ok(status)
+    }
+}
+
+async fn channel_handler(
+    cancellation_token: CancellationToken,
+    this_connection: ThisConnectionIdentifier,
+    target_connection: ConnectionIdentifier,
+    channel_identifier: ChannelIdentifier,
+    queue: ChannelControlQueueListener,
+    communication: impl Communication,
+) -> Result<ChannelHandlerStatus> {
+    debug!("Channel negotiated. Connecting to {target_connection} on channel {channel_identifier}");
+    let (reader, writer) = match create_channel(
+        this_connection,
+        target_connection.clone(),
+        channel_identifier,
+        communication,
+    )
+        .await
+    {
+        Ok((reader, writer)) => (reader, writer),
+        Err(e) => {
+            return Err(format!("Could not create channel to {target_connection}: {e:?}").into());
+        }
+    };
+
+    let mut handler = ChannelHandler::new(cancellation_token, queue, reader, writer);
+    handler.run().await
+}
+
+pub(super) fn create_channel_handler(
+    this_connection: ThisConnectionIdentifier,
+    target_connection: ConnectionIdentifier,
+    channel: ChannelIdentifier,
+    channel_cancellation_token: CancellationToken,
+    queue: ChannelControlQueueListener,
+    communication: impl Communication + 'static,
+    controller: NetworkingConnectionController,
+) {
+    tokio::spawn(
```

Comment:
We should make every top-level tokio task a struct in the future, pass it the data that it needs and give it a `run` fn or similar.